### ISSUE TITLE
glscopeclient hangs on Rigol MSO5000

### ISSUE THREAD

**Issue by pepijndevos** [azonenberg/scopehal#358]
When I configure the glscopeclient for my Rigol and press OK, it pops up a window if I want to kill the app or wait for it to respond, and then after a good few seconds the UI comes up.

It is extremely slow to load any data, showing like 3 WFMs, 0.05 WFM/s, and while the label of the sample rate matches what i see on the scope (2GSa/s, 4Mpts), the view is zoomed in to to 20.000 ns, while the scope shows 200us/div. One time I managed to zoom out, but generally the UI just hangs and the time scale vanishes completely when zooming too far.

I tried to trace the commands it sends:

```
$ ./src/glscopeclient/glscopeclient --debug --trace SCPISocketTransport
Connecting to SCPI oscilloscope at 169.254.67.109:5555
[SCPISocketTransport::SendCommand] Sending *IDN?
[SCPISocketTransport::ReadReply] Got RIGOL TECHNOLOGIES,MSO5074,MS5A221301338,00.01.03.00.01
[SCPISocketTransport::SendCommand] Sending SYST:OPT:STAT? RL2
[SCPISocketTransport::ReadReply] Got 0
[SCPISocketTransport::SendCommand] Sending WAV:FORM BYTE
[SCPISocketTransport::SendCommand] Sending WAV:MODE RAW
[SCPISocketTransport::SendCommand] Sending CHAN1:VERN ON
[SCPISocketTransport::SendCommand] Sending CHAN2:VERN ON
[SCPISocketTransport::SendCommand] Sending CHAN3:VERN ON
[SCPISocketTransport::SendCommand] Sending CHAN4:VERN ON
[SCPISocketTransport::SendCommand] Sending TIM:VERN ON
[SCPISocketTransport::SendCommand] Sending CHAN1:DISP ON
[SCPISocketTransport::SendCommand] Sending CHAN2:DISP ON
[SCPISocketTransport::SendCommand] Sending CHAN3:DISP ON
[SCPISocketTransport::SendCommand] Sending CHAN4:DISP ON
[SCPISocketTransport::SendCommand] Sending SING
[SCPISocketTransport::SendCommand] Sending TRIG:STAT?
[SCPISocketTransport::ReadReply] Got TD
[SCPISocketTransport::SendCommand]     Sending ACQ:MDEP?
[SCPISocketTransport::ReadReply]     Got 4.0000E+06
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN1
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9962E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN2
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9832E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN3
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9620E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN4
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9927E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending SING
[SCPISocketTransport::SendCommand]     Sending CHAN1:SCALE?
[SCPISocketTransport::ReadReply]     Got 1
[SCPISocketTransport::SendCommand]     Sending CHAN1:OFFS?
[SCPISocketTransport::ReadReply]     Got 0
[SCPISocketTransport::SendCommand]     Sending TRIG:EDGE:SOUR?
[SCPISocketTransport::ReadReply]     Got CHAN4
    Trigger source: CHAN4
[SCPISocketTransport::SendCommand]     Sending CHAN2:SCALE?
[SCPISocketTransport::ReadReply]     Got 1
[SCPISocketTransport::SendCommand]     Sending CHAN2:OFFS?
[SCPISocketTransport::ReadReply]     Got 0
[SCPISocketTransport::SendCommand]     Sending CHAN3:SCALE?
[SCPISocketTransport::ReadReply]     Got 1
[SCPISocketTransport::SendCommand]     Sending CHAN3:OFFS?
[SCPISocketTransport::ReadReply]     Got 0
[SCPISocketTransport::SendCommand]     Sending CHAN4:SCALE?
[SCPISocketTransport::ReadReply]     Got 1
[SCPISocketTransport::SendCommand]     Sending CHAN4:OFFS?
[SCPISocketTransport::ReadReply]     Got 0
[SCPISocketTransport::SendCommand]     Sending TRIG:EDGE:LEV?
[SCPISocketTransport::ReadReply]     Got 1.200000E0
[SCPISocketTransport::SendCommand] Sending CHAN1:DISP?
[SCPISocketTransport::ReadReply] Got 1
[SCPISocketTransport::SendCommand] Sending CHAN2:DISP?
[SCPISocketTransport::ReadReply] Got 1
[SCPISocketTransport::SendCommand] Sending CHAN3:DISP?
[SCPISocketTransport::ReadReply] Got 1
[SCPISocketTransport::SendCommand] Sending CHAN4:DISP?
[SCPISocketTransport::ReadReply] Got 1
[SCPISocketTransport::SendCommand] Sending TRIG:STAT?
[SCPISocketTransport::ReadReply] Got STOP
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN1
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9962E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN2
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9832E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN3
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9620E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN4
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9927E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
[SCPISocketTransport::SendCommand]     Sending SING
[SCPISocketTransport::SendCommand] Sending TRIG:STAT?
[SCPISocketTransport::ReadReply] Got TD
[SCPISocketTransport::SendCommand]     Sending WAV:SOUR CHAN1
[SCPISocketTransport::SendCommand]     Sending WAV:PRE?
[SCPISocketTransport::ReadReply]     Got 0,2,4000000,1,5.000000E-10,-1.000000E-3,0.000000,3.9962E-02,0,128
[SCPISocketTransport::SendCommand]     Sending WAV:STAR 1
[SCPISocketTransport::SendCommand]     Sending WAV:STOP 4000000
[SCPISocketTransport::SendCommand]     Sending WAV:DATA?
Shutting down view for waveform CHAN1
    Average frame interval: 3558.728 ms (0.28 FPS, 3 frames)
    ----------------------------------------------------------
    Task              | Total (ms) | Average (ms) | Percentage
    ----------------------------------------------------------
    Render            |     6254.8 |     2084.941 | 100.0 %
    Cairo             |       54.0 |       18.000 | 0.9 %
    Texture download  |        2.9 |        0.966 | 0.0 %
    Prepare           |      390.3 |      130.093 | 6.2 %
    Build index       |        0.0 |        0.012 | 0.0 %
    Geometry download |       28.3 |        9.450 | 0.5 %
    Composite         |        1.6 |        0.517 | 0.0 %
Killed
```
When I clicked the cross, it again showed the "this app is not responding" dialog and I had to kill it.

As suggested on IRC, I also ran it in gdb, to see where it deadlocks the UI.

All the threads:
```
* 1    Thread 0x7ffff5152a00 (LWP 10242) "glscopeclient" 0x00007ffff5f16620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
  2    Thread 0x7ffff4e23700 (LWP 10248) "gmain"         0x00007ffff5e3105f in poll () from /usr/lib/libc.so.6
  3    Thread 0x7fffeffff700 (LWP 10249) "gdbus"         0x00007ffff5e3105f in poll () from /usr/lib/libc.so.6
  5    Thread 0x7fffe7fff700 (LWP 10262) "ScopeThread"   0x00007ffff5f16b0c in recv () from /usr/lib/libpthread.so.0
```
All the stacktraces:
```
0x00007ffff5f16620 in __lll_lock_wait () from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007ffff5f16620 in __lll_lock_wait () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#1  0x00007ffff5f0ee81 in pthread_mutex_lock () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#2  0x00007ffff63b4367 in __gthread_mutex_lock(__gthread_mutex_t*) (__mutex=0x555555b26848)
    at /usr/include/c++/10.1.0/x86_64-pc-linux-gnu/bits/gthr-default.h:749
azonenberg/scopehal-apps#3  0x00007ffff63b43b7 in __gthread_recursive_mutex_lock(__gthread_recursive_mutex_t*) (__mutex=0x555555b26848)
    at /usr/include/c++/10.1.0/x86_64-pc-linux-gnu/bits/gthr-default.h:811
azonenberg/scopehal-apps#4  0x00007ffff63b43ec in std::recursive_mutex::lock() (this=0x555555b26848) at /usr/include/c++/10.1.0/mutex:106
azonenberg/scopehal-apps#5  0x00007ffff63b4448 in std::lock_guard<std::recursive_mutex>::lock_guard(std::recursive_mutex&) (this=0x7fffffffbce8, __m=...)
    at /usr/include/c++/10.1.0/bits/std_mutex.h:159
azonenberg/scopehal-apps#6  0x00007ffff63f54df in RigolOscilloscope::Start() (this=0x555555b26800)
    at /home/pepijn/code/scopehal-apps/lib/scopehal/RigolOscilloscope.cpp:659
azonenberg/scopehal-apps#7  0x0000555555680020 in OscilloscopeWindow::ArmTrigger(bool) (this=0x555555b61550, oneshot=false)
    at /home/pepijn/code/scopehal-apps/src/glscopeclient/OscilloscopeWindow.cpp:2203
azonenberg/scopehal-apps#8  0x000055555567fe1f in OscilloscopeWindow::OnStart() (this=0x555555b61550)
    at /home/pepijn/code/scopehal-apps/src/glscopeclient/OscilloscopeWindow.cpp:2165
azonenberg/scopehal-apps#9  0x000055555569df04 in sigc::bound_mem_functor0<void, OscilloscopeWindow>::operator()() const (this=0x5555558480d8)
    at /usr/include/sigc++-2.0/sigc++/functors/mem_fun.h:1991
azonenberg/scopehal-apps#10 0x000055555569bea6 in sigc::adaptor_functor<sigc::bound_mem_functor0<void, OscilloscopeWindow> >::operator()() const
    (this=0x5555558480d0) at /usr/include/sigc++-2.0/sigc++/adaptors/adaptor_trait.h:256
azonenberg/scopehal-apps#11 0x00005555556978d4 in sigc::internal::slot_call0<sigc::bound_mem_functor0<void, OscilloscopeWindow>, void>::call_it(sigc::internal::slot_rep*) (rep=0x5555558480a0) at /usr/include/sigc++-2.0/sigc++/functors/slot.h:136
azonenberg/scopehal-apps#12 0x00007ffff75802b9 in Glib::SignalProxyNormal::slot0_void_callback(_GObject*, void*) () at /usr/lib/libglibmm-2.4.so.1
azonenberg/scopehal-apps#13 0x00007ffff6793052 in g_closure_invoke () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#14 0x00007ffff67bb898 in  () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#15 0x00007ffff67af41d in g_signal_emit_valist () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#16 0x00007ffff67afba5 in g_signal_emit_by_name () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#17 0x00007ffff67af80a in g_signal_emit_valist () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#18 0x00007ffff67af980 in g_signal_emit () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#19 0x00007ffff6e59eb9 in  () at /usr/lib/libgtk-3.so.0
azonenberg/scopehal-apps#20 0x00007ffff6793052 in g_closure_invoke () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#21 0x00007ffff67bb2fc in  () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#22 0x00007ffff67af41d in g_signal_emit_valist () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#23 0x00007ffff67af980 in g_signal_emit () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#24 0x00007ffff6e59a41 in  () at /usr/lib/libgtk-3.so.0
azonenberg/scopehal-apps#25 0x00007ffff6e17b3b in  () at /usr/lib/libgtk-3.so.0
azonenberg/scopehal-apps#26 0x00007ffff67af80a in g_signal_emit_valist () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#27 0x00007ffff67af980 in g_signal_emit () at /usr/lib/libgobject-2.0.so.0
azonenberg/scopehal-apps#28 0x00007ffff6f2835d in  () at /usr/lib/libgtk-3.so.0
  
  #0  0x00007ffff5e3105f in poll () at /usr/lib/libc.so.6
azonenberg/scopehal-apps#1  0x00007ffff66f7168 in  () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#2  0x00007ffff66a8221 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#3  0x00007ffff66a8272 in  () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#4  0x00007ffff66d1511 in  () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#5  0x00007ffff5f0c422 in start_thread () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#6  0x00007ffff5e3bbf3 in clone () at /usr/lib/libc.so.6

#0  0x00007ffff5e3105f in poll () at /usr/lib/libc.so.6
azonenberg/scopehal-apps#1  0x00007ffff66f7168 in  () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#2  0x00007ffff66a7c03 in g_main_loop_run () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#3  0x00007ffff6a5f1a8 in  () at /usr/lib/libgio-2.0.so.0
azonenberg/scopehal-apps#4  0x00007ffff66d1511 in  () at /usr/lib/libglib-2.0.so.0
azonenberg/scopehal-apps#5  0x00007ffff5f0c422 in start_thread () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#6  0x00007ffff5e3bbf3 in clone () at /usr/lib/libc.so.6

#0  0x00007ffff5f16b0c in recv () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#1  0x00007ffff640fe43 in Socket::RecvLooped(unsigned char*, int) (this=0x555555b6b748, buf=0x7fffdc001270 "\n", len=4000001)
    at /home/pepijn/code/scopehal-apps/lib/xptools/Socket.cpp:311
azonenberg/scopehal-apps#2  0x00007ffff63a303d in SCPISocketTransport::ReadRawData(unsigned long, unsigned char*) (this=
    0x555555b6b740, len=4000001, buf=0x7fffdc001270 "\n") at /home/pepijn/code/scopehal-apps/lib/scopehal/SCPISocketTransport.cpp:135
azonenberg/scopehal-apps#3  0x00007ffff63f4d77 in RigolOscilloscope::AcquireData(bool) (this=0x555555b26800, toQueue=true)
    at /home/pepijn/code/scopehal-apps/lib/scopehal/RigolOscilloscope.cpp:599
azonenberg/scopehal-apps#4  0x00005555556f0ec7 in ScopeThread(Oscilloscope*) (scope=0x555555b26800)
    at /home/pepijn/code/scopehal-apps/src/glscopeclient/main.cpp:358
azonenberg/scopehal-apps#5  0x00005555556afd7c in std::__invoke_impl<void, void (*)(Oscilloscope*), Oscilloscope*>(std::__invoke_other, void (*&&)(Oscilloscope*), Oscilloscope*&&) (__f=@0x555555c40b40: 0x5555556f0caf <ScopeThread(Oscilloscope*)>) at /usr/include/c++/10.1.0/bits/invoke.h:60
azonenberg/scopehal-apps#6  0x00005555556afcd5 in std::__invoke<void (*)(Oscilloscope*), Oscilloscope*>(void (*&&)(Oscilloscope*), Oscilloscope*&&)
    (__fn=@0x555555c40b40: 0x5555556f0caf <ScopeThread(Oscilloscope*)>) at /usr/include/c++/10.1.0/bits/invoke.h:95
azonenberg/scopehal-apps#7  0x00005555556afc45 in std::thread::_Invoker<std::tuple<void (*)(Oscilloscope*), Oscilloscope*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (this=0x555555c40b38) at /usr/include/c++/10.1.0/thread:264
azonenberg/scopehal-apps#8  0x00005555556afbfe in std::thread::_Invoker<std::tuple<void (*)(Oscilloscope*), Oscilloscope*> >::operator()()
    (this=0x555555c40b38) at /usr/include/c++/10.1.0/thread:271
azonenberg/scopehal-apps#9  0x00005555556afbe2 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Oscilloscope*), Oscilloscope*> > >::_M_run() (this=0x555555c40b30) at /usr/include/c++/10.1.0/thread:215
azonenberg/scopehal-apps#10 0x00007ffff6196b74 in std::execute_native_thread_routine(void*) (__p=0x555555c40b30)
    at /build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
azonenberg/scopehal-apps#11 0x00007ffff5f0c422 in start_thread () at /usr/lib/libpthread.so.0
azonenberg/scopehal-apps#12 0x00007ffff5e3bbf3 in clone () at /usr/lib/libc.so.6
```

**Comment by azonenberg** [comment]
New root cause identified. Reopened and migrated to scopehal repo.

**Comment by bvernoux** [comment]
Feedback are welcome here it is rock stable with Windows7 64bits build with 0.94 WFM/s
![image](https://user-images.githubusercontent.com/432709/101055500-4091fb80-358a-11eb-96d6-b1a92c4cd14d.png)

**Comment by azonenberg** [comment]
Hopefully fixed for good in 6bfbfb4.
